scipy/optimize/_differentiable_functions.py

```diff
@@ -1,91 +1,119 @@
+from collections import namedtuple
 import numpy as np
 import scipy.sparse as sps
 from ._numdiff import approx_derivative, group_columns
 from ._hessian_update_strategy import HessianUpdateStrategy
 from scipy.sparse.linalg import LinearOperator
 from scipy._lib._array_api import array_namespace
 from scipy._lib import array_api_extra as xpx
+from scipy._lib._util import _ScalarFunctionWrapper
 
 
 FD_METHODS = ('2-point', '3-point', 'cs')
 
 
-def _wrapper_fun(fun, args=()):
-    ncalls = [0]
+class _GradWrapper:
```

Comment:
callable class to allow pickling, replacing a closure, which doesn't allow pickling.